tests/http_server/test_web_server.py

```diff
@@ -0,0 +1,154 @@
+#  Copyright 2023 Google LLC.
+#  Copyright (c) Microsoft Corporation.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+import http.server
+import itertools
+import socketserver
+import threading
+
+
+class _BaseServer(http.server.HTTPServer):
+    """Internal server that throws if timed out waiting for a request."""
+    def __init__(self, on_request):
+        """Starts the server."""
+        class _Handler(http.server.BaseHTTPRequestHandler):
+            """Internal handler that just asks the server to handle the request."""
+            def do_GET(self):
+                if self.path.endswith('favicon.ico'):
+                    self.send_error(404)
+                    return
+                on_request(self)
+
+            def do_POST(self):
+                on_request(self)
+
+            def handle(self):
+                http.server.BaseHTTPRequestHandler.handle(self)
+
+            def finish(self):
+                http.server.BaseHTTPRequestHandler.finish(self)
+
+            def log_message(self, *args, **kwargs):
+                """Overrides base class method to disable logging."""
+                pass
+
+        http.server.HTTPServer.__init__(self, ('127.0.0.1', 0), _Handler)
+
+        # TODO: implement HTTPS along with `ignoreHTTPSErrors` capability.
+        # if server_cert_and_key_path is not None:
+        #     self._is_https_enabled = True
+        #     self.socket = ssl.wrap_socket(self.socket,
+        #                                   certfile=server_cert_and_key_path,
+        #                                   server_side=True)
+        # else:
+        self._is_https_enabled = False
+
+    def handle_timeout(self):
+        """Overridden from SocketServer."""
+        raise RuntimeError('Timed out waiting for http request')
+
+    def get_url(self, host=None):
+        """Returns the base URL of the server."""
+        postfix = '://{}:{}'.format(host or '127.0.0.1', self.server_port)
+        if self._is_https_enabled:
+            return 'https' + postfix
+        return 'http' + postfix
+
+
+class _ThreadingServer(socketserver.ThreadingMixIn, _BaseServer):
+    """_BaseServer enhanced to handle multiple requests simultaneously"""
+    pass
+
+
+class StaticWebServer:
```

Comment:
meta: should we serve static files from filesystem or should we always inject HTML from the test? static files have an advantage that it is easy to open outside of the test runner.

Reply:
Serving content from parameters is easier to be integrated with the test, eg [here](https://github.com/GoogleChromeLabs/chromium-bidi/blob/4f1f659b7099af89265571940137faad68ea1d79/tests/browsing_context/test_create.py#L100).

Reply:
I mean that test files could be put as .html files on the file system as it is easier to open standalone. I am not sure what you mean by putting into JSON.

Reply:
> I mean that test files could be put as .html files on the file system as it is easier to open standalone. I am not sure what you mean by putting into JSON.

Static HTML pagers are already handled by `data:text/html,...` URLs. The question is how do you test redirects with the static HTML files?

Reply:
Note that data URLs are not 100% the same as static files. For example, you cannot assign an origin to the data URL and some other origin related things could be different.

Reply:
> Static HTML pagers are already handled by data:text/html,... URLs. The question is how do you test redirects with the static HTML files?

depends on what kind of redirect you want to do. In Chromium, for example, a redirect also a static file that instructs the server to redirect. Client side redirects are possible with static files.

Reply:
> static files have an advantage that it is easy to open outside of the test runner

It looks like I don't understand the proposal. The advantage of having the static files would be in the ability to access them by browser directly from file system omitting local server. But it is not a case when some complex setup is required, let's say response `301` with `Location`., for which some local server would be required anyways.

In WPT this setup totally makes sense, as apart of BiDi there are lots of tests checking across browsers complicated pages, eg some CSS features. In case of Mapper BiDi server is required to test complex workflows, not pages, eg OOPiF.

In both of cases there is an open question how to debug the tests manually. The easiest solution would be in adding something like this:
```
@pytest.mark.timeout(3600)
@pytest.mark.asyncio
async def test_run_server(websocket, context_id, url_200,
      url_301):
    print(url_301(url_200("MY CUSTOM PAGE")))

    while True:
        pass
```

Reply:
If we have static files, we could add a command to that the e2e server independently from the tests and then debug the test files like it's done in blink tests.

Reply:
> The advantage of having the static files would be in the ability to access them by browser directly from file system omitting local server. 

not omitting the server, but omitting the test runner/test code.

Reply:
E.g., `python -m http.server 8080 --bind 127.0.0.1 --directory /path/to/dir`. I believe it should be also possible to configure a directory for this server implementation to serve files from a directory.

Reply:
although it looks like SimpleHTTPRequestHandler is a different subclass in Python (that is the one that provides the directory serving).

Reply:
oh there is also CGIHTTPRequestHandler :)

Reply:
> Note CGI scripts run by the [CGIHTTPRequestHandler](https://docs.python.org/3/library/http.server.html#http.server.CGIHTTPRequestHandler) class cannot execute redirects (HTTP code 302), because code 200 (script output follows) is sent prior to execution of the CGI script. This pre-empts the status code.

Reply:
ah good to know, we could implement our own logic though if we need to. For reference, here is the "static" file redirects in web tests https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/http/tests/navigation/resources/redirect302.pl;drc=77578ccb4082ae20a9326d9e673225f1189ebb63 you can debug them by running `third_party/blink/tools/run_blink_httpd.py` without the need to run the tests.

Reply:
Alternatively to [manual run from test](https://github.com/GoogleChromeLabs/chromium-bidi/pull/1359#discussion_r1341243232), we can put in files JSON with `code`, `headers` and `content`. I guess it's what you are suggesting.

Conventional file-based approach requires slightly more harness around, and would require a separate script user will have to run manually in order to have a server running for debugging purposes. From the other side, it would make debugging more easy and consistent.

I'm slightly inclined to start with the 1st approach, and fall back to the 2nd if we find it's not ergonomic.

Reply:
putting them to JSON still requires editing test files (HTML, CSS, images, video/audio potentially more formats) within JSON so not much difference compared to editing in Python literals. But yes it would allow running it outside of the test runner. Alternatively, we could, for example, add a metadata data file `.http` to provider additional HTTP headers for static files. That would require checking if there is a metadata file for any given static file.